### ISSUE TITLE
Broken with floats if locale decimal_point is not '.'

### ISSUE THREAD

**Issue by piru** [troldal/OpenXLSX#190]
OpenXLSX fails when used with locales that use `decimal_point` other than `.`. This happens because of checking for hardcoded decimal separator at https://github.com/troldal/OpenXLSX/blob/29034465e569675f1677157b65b8263ccb1452f0/OpenXLSX/sources/XLCellValue.cpp#L249

**Comment by CEXT-Dan** [comment]
It’s my understanding that this is by design. If it were by locale, then other locales may see an invalid file format.

**Comment by piru** [comment]
OpenXLSX uses locale aware routines elsewhere to format floats. Hence it itself generates invalid formats. Either you consistently obey locale or don't. You can't mix.

**Comment by CEXT-Dan** [comment]
I don't think this is the case in the raw data, I believe this is stored in the STYLE, otherwise, if you were to send me a file to test, my tests would fail.

**Comment by piru** [comment]
Here are steps to reproduce:

1. Apply this patch:
```diff --git a/Examples/Demo1.cpp b/Examples/Demo1.cpp
index 44d6eee..102b571 100644
--- a/Examples/Demo1.cpp
+++ b/Examples/Demo1.cpp
@@ -7,6 +7,8 @@ using namespace OpenXLSX;

 int main()
 {
+    setlocale(LC_ALL, "");
+
     cout << "********************************************************************************\n";
     cout << "DEMO PROGRAM #01: Basic Usage\n";
     cout << "********************************************************************************\n";
```
2. Rebuild
3. `LC_NUMERIC=fi_FI.UTF-8 output/Demo1` (Note: Any locale that has non `.` `decimal_point` will do, `fi_FI.UTF-8` is just one of them)

Result:
```
********************************************************************************
DEMO PROGRAM #01: Basic Usage
********************************************************************************
terminate called after throwing an instance of 'OpenXLSX::XLValueTypeError'
  what():  XLCellValue object does not contain the requested type.
Aborted
```

**Comment by CEXT-Dan** [comment]
I see, IMHO, the issue is in PUGI__SNPRINTF is using your locale when setting the data.
The raw data itself should always be “.”

**Comment by piru** [comment]
This is a bit messy to fix cleanly on all platforms. On `POSIX 2008` conforming system this is somewhat ok, as you can `uselocale` the `C` locale and use it with `snprintf_l` and `wcstod_l`/`strtod_l`.

Here are some discussion: https://stackoverflow.com/questions/13919817/how-to-parse-numbers-like-3-14-with-scanf-when-locale-expects-3-14

**Comment by aral-matrix** [comment]
Hi @piru - I am going through the old issues backlog at the moment, and I can not reproduce your example - I did however fix a bug today that caused the above exception, which was that if a clean floating point value was stored in a date value, a XLCellValue::get<double> would fail because it would read the cell as an integer.

I believe that the exception you saw might have been thrown because:
```
    wks.cell("A1").value() = 3.14159265358979323846; // this stores a float
```
As mentioned by @CEXT-Dan, pugixml then potentially that with a different decimal & something undefined goes on (hard to say without having the exact reproducible scenario).
```
    cout << "Cell A1: (" << A1.typeAsString() << ") " << A1.get<double>() << endl;
```
When reading from A1, OpenXLSX only gets an integer value from the underlying XML (due to whatever pugixml did) and std::get< double > throws.

The current development-aral branch has implicit number conversion to float for such a scenario, so you should see "rounding" if the decimal separator is still inconsistent - e.g. an output of "3" instead of ```3.14...```.

Either way, OOXML should only ever store ```.``` as decimal separator, the locale is display-only for your flavor of Office. Please do not use OpenXLSX with any ```setlocale``` calls. The correct way to achieve formatting in the display(!) of numbers in your XLSX file is described here:
https://github.com/troldal/OpenXLSX/issues/181

Closing this issue as resolved. Please open a new one if there is a reasonable feature request that could help achieve generic functionality.
